### PR TITLE
Lecture07perf

### PR DESCRIPTION
**Pull request by timspainUCL**
Add the concurrency notes relating to the classwork.

### CODE REVIEW

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+### The set up
+Five hungry yet contemplative philosophers are seated around a table. In front of each of them is a bowl of noodles. In between each of them is a single chopstick. The philosophers are holding one of their famous _symposia_, a party for eathing and thinking and drinking. Each philosopher will alternate between thinking and eating. To eat, a philosopher needs to have hold of _both_ chopsticks. Crucially, this means that not all of the philosophers can eat at the same time.
+
+ ![Dining philosophers](dining_philosophers2.png)
+
+A fuller explanation of the Dining Philosphers problem can be found on [Wikipedia](https://en.wikipedia.org/wiki/Dining_philosophers_problem).
+
+## Deadlock
+It is easy for the philosophers to get into what is called a deadlocked state. In the analogy, each philosopher is determined to eat, and picks up the chopstick to their right, and is unwilling to put it down until after they have eaten. They are each left holding exactly one chopstick each, and are also left hungry, each unwilling to put down the chopstick the have hold of.
```
Comment (jeydobson) [inline 07PerformanceProgramming/Sec04ConcurrencyProblems.md:24]: Typo: eathing

Comment (jeydobson) [reply 07PerformanceProgramming/Sec04ConcurrencyProblems.md:24]: also suggest to remove drinking as not referenced in example

Comment (jeydobson) [reply 07PerformanceProgramming/Sec04ConcurrencyProblems.md:24]: typo: the have hold of

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+A simple solution to a deadlock is to implement a try-and-wait algorithm.
+
+For the philosophers, this means trying to eat using both chopsticks. If they are, great!, they get to eat. If not, then put down any chopsticks you may have in hand, wait a period of time and repeat.
+
+For the computer system, the procedure is very similar. Try to acquire the necessary resources, use them if available, otherwise free them for use by others and wait for some period of time before trying again.
+
+In both cases, this breaks the time symmetry of the problem. No longer are all the workers or philosphers in the same state for ever more, something is changing, and work gets done.
+
+## Livelock
+Or does it? The try-and-wait algorithm might end up with all of the philosophers picking up on chopstick simultaneously, realising the other is not available, and putting it back dowmn again without eating. They all wait for the same length of time, and try again.
```
Comment (jeydobson) [inline 07PerformanceProgramming/Sec04ConcurrencyProblems.md:38]: typos: "dowmn" "picking up on"

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+## Race condition
+Another problem that may occur in concurrent programs is that of race conditions. These occur when one thread checks if a resource is available, is interrupted by another that makes that resource unavailable, resumes and is unable to complete the intended action. The effect of this can be anything from starving one of the workers (as above) to causing the entire program or system to crash.
+
+In the case of our philosophers, P3, wants to eat. They check that chopsticks C3 and C4 are available, and picks up C3. In the meantime, P4 has grabbed C4. What happens when P3 tries to pick up the chopstick that is no longer there? In the case of the philosophers, there would no doubt be a full and frank exchange of views. 
+
+### Waiter/arbitrator
+One solution that race conditions is to use an arbitrator for access to the contended resources. In the case of the dining philosophers, this can be thought of as a waiter whom the philosophers ask permission to pick up both their chopsticks. One one is allowed to pick up chopsticks at a time, and the waiter will not allow another to attempt to do so until the previous philosopher has their chopsticks firmly in hand.
```
Comment (jeydobson) [inline 07PerformanceProgramming/Sec04ConcurrencyProblems.md:58]: Typo: "One one" -> "Only one"

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+The waiter can also resolve the issue of resource starvation by ensuring that not philosopher is allowed to go too long without getting to pick up chopsticks and eat.
```
Comment (jeydobson) [inline 07PerformanceProgramming/Sec04ConcurrencyProblems.md:60]: typo: "not" -> "no"

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+A mutex is an object that is shared between threads which allows only one thread to pass a certain point at any one time. The name is short for *mut*ual *ex*clusion, and might also be known as a lock.
+
+A mutex protects access to what is known as a critical section to only one worker thread at a time. In the case of the dining philosophers, this critical section is the action of picking up chopsticks. The worker thread attempts to access the critical section. If no other worker is within, the mutex will allow the worker thread to proceed, perform the actions of the critical section and crucially to release the mutex lock when it is finished. This allows any other thread that might be waiting to proceed into the critical section.
+
+### `std::mutex`
+Introduced in C++11.
+
+The `std::mutex` class provides a minimal, portable interface to the underlying operating system mutexes. It can be included in your code using the `#include <mutex>` standard header inclusion. The base C++ version provides the following member functions:
+* `mutex( )`
+ * default constructor
```
Comment (jeydobson) [inline 07PerformanceProgramming/Sec04ConcurrencyProblems.md:76]: Check indentation? On the preview all indents at same level but maybe that doesn't properly preview it

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+### `std::atomic`
+This is a template that ensures that access to a variable is atomic. It can be included in your code using the `#include <atomic>` standard header inclusion.
+
+The header also includes atmoic versions of the built-in `bool` and integer (`char`, `int`, `long`) types enforcing atomic access, arithmetic and incrementing.
```
Comment (jeydobson) [inline 07PerformanceProgramming/Sec04ConcurrencyProblems.md:94]: Typo: "atmoic"

File: 07PerformanceProgramming/index.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+---
+title: "Lecture 7: Performance Programming"
+---
+
+## Lecture 7: Overview
+
+### Today's Lesson
+
+- Performance Programming
+ - Why?
```
Comment (jeydobson) [inline 07PerformanceProgramming/index.md:10]: Check indentation? Showing up as single level of indentation in the preview.

**Review by jeydobson — APPROVED** [review]
Looks good. I pointed out a few typos so ok to merge after you taken a look at those.

File: 07PerformanceProgramming/index.md (hunk 7 above)
Comment (timspainUCL) [reply 07PerformanceProgramming/index.md]: Done

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md (hunk 6 above)
Comment (timspainUCL) [reply 07PerformanceProgramming/Sec04ConcurrencyProblems.md]: The curse of the atomspheric physicist

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md (hunk 5 above)
Comment (timspainUCL) [reply 07PerformanceProgramming/Sec04ConcurrencyProblems.md]: Checked and corrected

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md (hunk 4 above)
Comment (timspainUCL) [reply 07PerformanceProgramming/Sec04ConcurrencyProblems.md]: Done

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md (hunk 3 above)
Comment (timspainUCL) [reply 07PerformanceProgramming/Sec04ConcurrencyProblems.md]: Donly

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md (hunk 2 above)
Comment (timspainUCL) [reply 07PerformanceProgramming/Sec04ConcurrencyProblems.md]: Done

File: 07PerformanceProgramming/Sec04ConcurrencyProblems.md (hunk 1 above)
Comment (timspainUCL) [reply 07PerformanceProgramming/Sec04ConcurrencyProblems.md]: Symposium specifically means drinking together,  but point taken.

**Pull request merged**